### ISSUE TITLE
What is the best way to insert 100k messages per second?

### ISSUE THREAD

**Issue by hpham04** [OpenTSDB/opentsdb#1414]
Let’s say we have 100k devices, each of them sends a message per second. It means we need to be able to insert 100k in to opentsdb in one second
Calling rest api seems to be not a good approach because we need to open 100k connections every second

**Comment by jmritz** [comment]
The best thing you could do is write a plugin.  You can read about TSDB plugins [here](http://opentsdb.net/docs/build/html/user_guide/plugins.html).  Specifically the RPC plugin.

**Comment by johnseekins** [comment]
It's also worth considering that the docs recommend persistent connections for the rest api. So you should be opening 100k connections _once_, then maintaining them over time.

**Comment by hpham04** [comment]
As far as I know, http 1/1 uses persistent connections by default and opentsdb persistent connection default value is also on, so there is nothing to do
Please correct me if I’m wrong 
Thank you so much

**Comment by GreatSnoopy** [comment]
A connection is a two ended thing. If your  devices wake up, connect and send some data and then disconnect, it does not matter that opentsdb allows a connection to be kept open.
As other said, put something in front of your system that is able to accept data item by item and then aggregates data in batches and writes them into tsdb. In addition, having a message queue in the blend will greatly help to normalize peaks.

I would also add that your question is a little bit ambiguous, as it does not makes clear enough what exactly your challenge is: 
- tsdb scalability - what resources you need to be able to do 100k?
- ingestion performance (how should i write data to get to 100k?)

I would assume you did some initial evaluation - what was the wall you hit?

**Comment by johnewing1** [comment]
I'd agree with the above, putting a queue in the way is really important.

We run a resonably large tsdb install doing 500-600k writes per second and have a small web service that implements the opentsdb put api and writes the data points to kafka. We've written our own version of this 
https://github.com/OpenTSDB/opentsdb-rpc-kafka to consume from kafka and persist the points to the store.

You will also need to do a bit of tuning for that type of write rate. Make sure you pre split the tables if you haven't already.  Look at salting http://opentsdb.net/docs/build/html/user_guide/writing/index.html#salting to make sure you get a good spread of write load across the region servers.

**Comment by johnseekins** [comment]
Put something in the way and write/find a tool that will read from that something and write to OpenTSDB using persistent connections. In our case, we've used a Flink job, but anything that can read from a queue and write to OpenTSDB would work.
Looks like Logstash has this ability, for example.

**Comment by manolama** [comment]
The kafka RPC plugin is the closest as @mancubus77 mentioned and I'll try to get that fixed up as soon as I cut 2.4 here. With a single TSDB on a decent machine (8 cores, 16GB ram) and lots of sources writing you can easily get around 20k with the HTTP API, around 30k with the Telnet API and well over 100k with the Java API.

**Comment by johann8384** [comment]
I don't think you should need your own plugin to write 100k/s. You should be able to write into the millions of datapoints per second on the telnet interface. Writing a plugin and building out a Kafka environment just to insert 100k dps seems a little overkill.
